src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/BasicClientFragment.kt

```diff
@@ -0,0 +1,147 @@
+    basicClusterFabricIdEd.setText(testFabricId)
+    basicClusterDeviceIdEd.setText(DeviceIdUtil.getLastDeviceId(requireContext()).toString())
+  }
+
+  inner class ChipControllerCallback : GenericChipDeviceListener() {
+    override fun onConnectDeviceComplete() {}
+
+    override fun onCommissioningComplete(nodeId: Long, errorCode: Int) {
+      Log.d(TAG, "onCommissioningComplete for nodeId $nodeId: $errorCode")
+    }
+
+    override fun onNotifyChipConnectionClosed() {
+      Log.d(TAG, "onNotifyChipConnectionClosed")
+    }
+
+    override fun onCloseBleComplete() {
+      Log.d(TAG, "onCloseBleComplete")
+    }
+
+    override fun onError(error: Throwable?) {
+      Log.d(TAG, "onError: $error")
+    }
+  }
+
+  override fun onStop() {
+    super.onStop()
+    scope.cancel()
+  }
+
+  private fun updateAddressClick() {
+    try {
+      deviceController.updateDevice(
+              basicClusterFabricIdEd.text.toString().toULong().toLong(),
+              basicClusterDeviceIdEd.text.toString().toULong().toLong()
+      )
+      showMessage("Address update started")
+    } catch (ex: Exception) {
+      showMessage("Address update failed: $ex")
+    }
+  }
+
+  private suspend fun sendReadProductNameCommandClick() {
+    getBasicClusterForDevice().readProductNameAttribute(object : ChipClusters.CharStringAttributeCallback {
+      override fun onSuccess(value: String) {
+        showMessage("Read ProductName command success: value: $value")
+      }
+
+      override fun onError(ex: Exception) {
+        showMessage("Read ProductName command failure $ex")
+        Log.e(TAG, "Read ProductName command failure", ex)
+      }
+    })
+  }
+
+  private suspend fun sendReadUserLabelCommandClick() {
+    getBasicClusterForDevice().readUserLabelAttribute(object : ChipClusters.CharStringAttributeCallback {
+      override fun onSuccess(value: String) {
+        showMessage("Read UserLabel command success: value: $value")
+      }
+
+      override fun onError(ex: Exception) {
+        showMessage("Read UserLabel command failure $ex")
+        Log.e(TAG, "Read UserLabel command failure", ex)
+      }
+    })
+  }
+
+  private suspend fun sendWriteUserLabelCommandClick() {
+    getBasicClusterForDevice().writeUserLabelAttribute(object : ChipClusters.DefaultClusterCallback {
+      override fun onSuccess() {
+        showMessage("Write UserLabel command success")
+      }
+
+      override fun onError(ex: Exception) {
+        showMessage("Write UserLabel command failure $ex")
+        Log.e(TAG, "Write UserLabel command failure", ex)
+      }
+
+    }, userLabelEd.text.toString())
+  }
+
+  private fun showMessage(msg: String) {
+    requireActivity().runOnUiThread {
+      basicClusterCommandStatus.text = msg
+    }
+  }
+
+  private suspend fun getBasicClusterForDevice(): BasicCluster {
+    return BasicCluster(
+            ChipClient.getConnectedDevicePointer(requireContext(), basicClusterDeviceIdEd.text.toString().toLong()), 0
```

Comment:
Same as above, indentation should be two spaces past the beginning of `return`

Reply:
Thank you for feedback. fixed to two spaces